packages/persist/lib/server.ts

```diff
@@ -0,0 +1,82 @@
+import express from 'express';
+import type { Request, Response, NextFunction } from 'express';
+import { validateRequest } from 'zod-express';
+import { z } from 'zod';
+import persistController from './controllers/persist.controller.js';
+
+export const server = express();
+server.use(express.json());
+//TODO: tracing
+
+server.get('/health', (_req: Request, res: Response) => {
+    res.json({ status: 'ok' });
+});
+server.put(
+    '/sync/:syncId',
+    validateRequest({
+        params: z.object({
+            syncId: z.string()
+        }),
+        body: z.object({
+            lastSyncDate: z
+                .string()
+                .datetime()
+                .transform((value) => new Date(value))
+                .pipe(z.date()) as unknown as z.ZodDate
+        })
+    }),
+    persistController.saveLastSyncDate
+);
+
+server.post(
+    '/environment/:environmentId/log',
+    validateRequest({
+        params: z.object({
+            environmentId: z.string().transform(Number).pipe(z.number().int().positive()) as unknown as z.ZodNumber
+        }),
+        body: z.object({
+            activityLogId: z.number(),
+            level: z.enum(['info', 'debug', 'error', 'warn', 'http', 'verbose', 'silly']),
+            msg: z.string()
+        })
+    }),
+    persistController.saveActivityLog
+);
+
+const validateRecordsRequest = validateRequest({
+    params: z.object({
+        environmentId: z.string().transform(Number).pipe(z.number().int().positive()) as unknown as z.ZodNumber,
+        connectionId: z.string(),
+        syncId: z.string(),
+        syncJobId: z.string().transform(Number).pipe(z.number().int().positive()) as unknown as z.ZodNumber
+    }),
+    body: z.object({
+        model: z.string(),
+        records: z.any().array().nonempty(),
+        providerConfigKey: z.string(),
+        nangoConnectionId: z.number(),
+        activityLogId: z.number(),
+        lastSyncDate: z
+            .string()
+            .datetime()
+            .transform((value) => new Date(value))
+            .pipe(z.date()) as unknown as z.ZodDate,
+        trackDeletes: z.boolean()
+    })
+});
+server.post('/environment/:environmentId/connection/:connectionId/sync/:syncId/job/:syncJobId/records', validateRecordsRequest, persistController.saveRecords);
+server.delete(
+    '/environment/:environmentId/connection/:connectionId/sync/:syncId/job/:syncJobId/records',
+    validateRecordsRequest,
+    persistController.deleteRecords
+);
+server.put('/environment/:environmentId/connection/:connectionId/sync/:syncId/job/:syncJobId/records', validateRecordsRequest, persistController.updateRecords);
+
+// error handling middleware
+server.use((err: Error, _req: Request, res: Response, next: NextFunction) => {
+    if (err) {
+        res.status(500).json({ error: err.message });
+    } else {
+        next();
+    }
+});
```

Comment:
do we want some sort of authorization mechanism to make sure requests are legit and actually made from a runner?

Reply:
If the persist service is a private service then I don't think that is necessary yet.